Confidential-Storage-Arch.md

```diff
@@ -0,0 +1,106 @@
+- virtio-scsi devices
+- virtio-pmem devices  
+- nvme devices
+- virtio-fs devices
+- virtio-9p devices
+- network attached filesystems, such as NFS, CEPHFS etc. There's a special case for network attached filesystems, it may
+  be trusted filesystem if it's connected to a trusted server through secure connections.
+- software services providing storage data, for example an image cache service integrated into the image service.  
+
+The storage subsystem may be composed up with multiple layers, so we may enforce confidentiality at different positions.
+Here confidentiality enforcement means decryption and/or integrity verification.
+
+### Enforce Confidentiality at The Data Source Layer
+The confidentiality may be enforced at the data source layer, such as at the service agents or the NFS clients.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Data Source Layer](images/storage-enforce-confidentiality-at-source.png)
+
+### Enforce Confidentiality at The Block Device Layer
+The confidentiality may also be enforced at the block device layer, such as by using the dm-crypt/dm-verity.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Block Device Layer](images/storage-enforce-confidentiality-at-blockdev.png)
+
+### Enforce Confidentiality at The FileSystem Layer
+The confidentiality may also be enforced at the filesystem layer, such as by using the to-be-invented filesystems.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Filesystem Layer](images/storage-enforce-confidentiality-at-fs.png)
+
+### Composition of Confidential Filesystem
+Once those confidential filesystems are ready, they could be composed up as normal filesystems in anyway, such as by
+mount, bind mount, overlayfs, chroot or pivot etc.
+
+## Ownership and Life Management of Storage Devices
+With traditional Kata Containers, it aims to be compatible with runC and all filesystems are prepared on host, 
+the Kata runtime just pass those storage resources into the guest. With confidential containers, that model doesn't
+work anymore, it needs to prepare filesystems for containers within the vm.
+
+In addition to normal container associated volume, a new concept of "pod built-in ephemeral volume/storage device" 
+is introduced. The pod built-in ephemeral volume  which will be used to:
+- host guest os temporary data 
+- provide ephemeral volumes for containers
+- store downloaded container images
+- store rw layer of container rootfs
+
+The lifecycle of the pod built-in ephemeral volume will be managed by the orchestration system or by the host.
+It may be associated with the pod/container, and get destroyed when deleting the pod/container. Or it may be reused
+for different pods/containers to improve performance.
+
+### Overall Architecture
+To summarize, we hope to build a layered confidential storage architecture for confidential containers as below:
+
+![Layered Confidential Storage Architecture](images/storage-layerd-arch.png)
```

Comment:
How about confidential file system on an already confidential block device? What I mean that will confidential file system implementation divides into two cases: 
1  Confidential file system builds on confidential block device. Then it can rely on confidential block device.
2  Confidential file system builds on ro stoarge device as shown in your proposed figure.

Reply:
Confidential fs over confidential blockdev should be an valid configuration too:)

Reply:
> Confidential fs over confidential blockdev should be an valid configuration too:)

So the diagram may need be slightly revised to cover this case.